File: src/corral/background_tasks/git_poller.py
"""Background poller that queries git state for live corral agents."""

from __future__ import annotations

import asyncio
import logging
import os
from pathlib import Path
from typing import Any

from corral.tools.session_manager import discover_corral_agents
from corral.tools.tmux_manager import _find_pane
from corral.store import CorralStore
from corral.tools.utils import run_cmd, HISTORY_PATH

log = logging.getLogger(__name__)


class GitPoller:
    """Periodically polls git branch/commit info for live agents and stores snapshots."""

    def __init__(self, store: CorralStore) -> None:
        self._store = store

    async def run_forever(self, interval: float = 30) -> None:
        while True:
            try:
                await self.poll_once()
            except Exception:
                log.exception("GitPoller error")
            await asyncio.sleep(interval)

    async def poll_once(self) -> dict[str, int]:
        agents = await discover_corral_agents()
        polled = 0

        # Group agents by working directory so we query git once per directory
        # but store a snapshot for every session in that directory.
        dir_to_agents: dict[str, list[dict[str, Any]]] = {}
        for agent in agents:
            session_id = agent.get("session_id")
            pane = await _find_pane(
                agent["agent_name"], agent["agent_type"], session_id=session_id,
            )
            if not pane:
                continue
            workdir = pane.get("current_path", "")
            if not workdir:
                continue
            dir_to_agents.setdefault(workdir, []).append(agent)

        for workdir, dir_agents in dir_to_agents.items():
            try:
                git_info = await self._query_git(workdir)
                if not git_info:
                    continue
                changed_files = await self._query_changed_files(workdir)
                # Store a snapshot for each session in this directory
                for agent in dir_agents:
                    await self._store.upsert_git_snapshot(
                        agent_name=agent["agent_name"],
                        agent_type=agent["agent_type"],
                        working_directory=workdir,
                        branch=git_info["branch"],
                        commit_hash=git_info["commit_hash"],
                        commit_subject=git_info["commit_subject"],
                        commit_timestamp=git_info["commit_timestamp"],
                        session_id=agent.get("session_id"),
                        remote_url=git_info.get("remote_url"),
                    )
                    await self._store.replace_changed_files(
                        agent_name=agent["agent_name"],
                        working_directory=workdir,
                        files=changed_files,
                        session_id=agent.get("session_id"),
                    )
                    polled += 1
            except Exception:
                log.exception("GitPoller error for dir %s", workdir)
        return {"polled": polled}

    async def _query_git(self, workdir: str) -> dict[str, str] | None:
        """Query git for current branch and latest commit in a working directory."""
        try:
            # Get branch name
            rc, stdout, _ = await run_cmd(
                "git", "-C", workdir, "rev-parse", "--abbrev-ref", "HEAD", timeout=5.0
            )
            if rc != 0:
                return None
            branch = stdout

            # Get latest commit: hash|subject|timestamp
            rc, stdout, _ = await run_cmd(
                "git", "-C", workdir, "log", "-1", "--format=%H|%s|%aI", timeout=5.0
            )
            if rc != 0:
                return None
            parts = stdout.split("|", 2)
            if len(parts) < 3:
                return None

            # Get remote URL (best-effort)
            remote_url = None
            try:
                rc, stdout, _ = await run_cmd(
                    "git", "-C", workdir, "remote", "get-url", "origin", timeout=5.0
                )
                if rc == 0:
                    remote_url = stdout or None
            except Exception:
                pass

            return {
                "branch": branch,
                "commit_hash": parts[0],
                "commit_subject": parts[1],
                "commit_timestamp": parts[2],
                "remote_url": remote_url,
            }
        except (asyncio.TimeoutError, OSError, FileNotFoundError):
            return None

    async def _get_diff_base(self, workdir: str) -> str:
        """Return the base ref to diff against.

        On a feature branch: merge-base with main/master (shows all branch work).
        On the default branch (or merge-base fails): HEAD (shows changes since last commit).
        """
        # Detect current branch
        rc, branch, _ = await run_cmd(
            "git", "-C", workdir, "rev-parse", "--abbrev-ref", "HEAD", timeout=5.0,
        )
        current_branch = branch.strip() if rc == 0 else ""

        # If we're not on a default branch, try to find the merge-base
        if current_branch not in ("main", "master", "HEAD", ""):
            for base_branch in ("main", "master"):
                rc, stdout, _ = await run_cmd(
                    "git", "-C", workdir, "merge-base", base_branch, "HEAD", timeout=5.0,
                )
                if rc == 0 and stdout:
                    return stdout.strip()

        # On the default branch or merge-base unavailable — diff against HEAD
        # (shows only uncommitted changes since the last commit)
        return "HEAD"

    async def _get_base_timestamp(self, workdir: str, base_ref: str) -> float:
        """Get the unix timestamp of the base ref commit.

        Used to filter out untracked files that existed before the base point.
        """
        rc, stdout, _ = await run_cmd(
            "git", "-C", workdir, "log", "-1", "--format=%ct", base_ref, timeout=5.0,
        )
        if rc == 0 and stdout:
            try:
                return float(stdout.strip())
            except ValueError:
                pass
        return 0.0

    async def _query_changed_files(self, workdir: str) -> list[dict[str, Any]]:
        """Query git for files changed relative to a smart base ref.

        On a feature branch: shows all changes since the branch diverged from main.
        On main/master: shows only changes since the last commit.
        Untracked files are only included if they were created after the base
        commit timestamp, so pre-existing untracked files don't pollute the list.
        """
        file_map: dict[str, dict[str, Any]] = {}

        try:
            base = await self._get_diff_base(workdir)
            base_ts = await self._get_base_timestamp(workdir, base)

            # Diff from base to working tree — captures committed (on branch) +
            # staged + unstaged changes in one shot.
            rc, stdout, _ = await run_cmd(
                "git", "-C", workdir, "diff", base, "--numstat", timeout=5.0,
            )
            if rc == 0 and stdout:
                for line in stdout.strip().split("\n"):
                    if not line.strip():
                        continue
                    parts = line.split("\t", 2)
                    if len(parts) < 3:
                        continue
                    added, removed, filepath = parts
                    a = int(added) if added != "-" else 0
                    d = int(removed) if removed != "-" else 0
                    file_map[filepath] = {"filepath": filepath, "additions": a, "deletions": d, "status": "M"}

            # Untracked files don't appear in git diff, so pick up '??' entries
            # from git status — but only if they were created after the base commit.
            rc, stdout, _ = await run_cmd(
                "git", "-C", workdir, "status", "--porcelain", "--untracked-files=all", timeout=5.0,
            )
            if rc == 0 and stdout:
                for line in stdout.strip().split("\n"):
                    if len(line) < 4:
                        continue
                    status_code = line[:2].strip()
                    filepath = line[3:]
                    if " -> " in filepath:
                        filepath = filepath.split(" -> ", 1)[1]
                    if status_code == "??":
                        if filepath not in file_map:
                            # Only include untracked files created after the base commit
                            full_path = os.path.join(workdir, filepath)
                            try:
                                mtime = os.path.getmtime(full_path)
                            except OSError:
                                continue
                            if mtime < base_ts:
                                continue
                            file_map[filepath] = {"filepath": filepath, "additions": 0, "deletions": 0, "status": "??"}
                    elif filepath in file_map:
                        file_map[filepath]["status"] = status_code

        except (asyncio.TimeoutError, OSError, FileNotFoundError):
            pass

        return list(file_map.values())

File: coral-go/internal/background/git_poller.go
// Package background provides long-running background services for Coral.
package background

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/cdknorow/coral/internal/executil"
	"github.com/cdknorow/coral/internal/gitutil"
	"github.com/cdknorow/coral/internal/store"
)

// GitPoller periodically polls git branch/commit info for live agents.
type GitPoller struct {
	store      *store.GitStore
	runtime    AgentRuntime
	interval   time.Duration
	logger     *slog.Logger
	discoverFn func(ctx context.Context) ([]AgentInfo, error)
	prCache    map[string]prCacheEntry // keyed by "remote_url::branch"
}

type prCacheEntry struct {
	prNumber  int
	expiresAt time.Time
}

// AgentInfo holds minimal agent metadata needed by background services.
type AgentInfo struct {
	AgentName        string
	AgentType        string
	SessionID        string
	WorkingDirectory string
	DisplayName      string // role name, used as stable board subscriber_id
}

// NewGitPoller creates a new GitPoller.
func NewGitPoller(gitStore *store.GitStore, runtime AgentRuntime, interval time.Duration) *GitPoller {
	return &GitPoller{
		store:    gitStore,
		runtime:  runtime,
		interval: interval,
		logger:   slog.Default().With("service", "git_poller"),
		prCache:  make(map[string]prCacheEntry),
	}
}

// SetDiscoverFn sets a custom agent discovery function (for dependency injection).
func (p *GitPoller) SetDiscoverFn(fn func(ctx context.Context) ([]AgentInfo, error)) {
	p.discoverFn = fn
}

// Run starts the polling loop. Blocks until context is cancelled.
func (p *GitPoller) Run(ctx context.Context) error {
	ticker := time.NewTicker(p.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			if err := p.PollOnce(ctx); err != nil {
				p.logger.Error("poll error", "error", err)
			}
		}
	}
}

// PollOnce performs a single git polling pass for all live agents.
func (p *GitPoller) PollOnce(ctx context.Context) error {
	pollStart := time.Now()

	agents, err := p.discoverAgents(ctx)
	if err != nil {
		return fmt.Errorf("discover agents: %w", err)
	}

	// Group agents by working directory to minimize git calls
	dirToAgents := make(map[string][]AgentInfo)
	for _, agent := range agents {
		if agent.WorkingDirectory == "" {
			continue
		}
		dirToAgents[agent.WorkingDirectory] = append(dirToAgents[agent.WorkingDirectory], agent)
	}

	p.logger.Debug("git poll starting", "dirs", len(dirToAgents), "agents", len(agents))

	for workdir, dirAgents := range dirToAgents {
		dirStart := time.Now()

		resolveStart := time.Now()
		workdir = gitutil.ResolveGitRoot(ctx, workdir)
		p.logger.Debug("git resolve root", "workdir", workdir, "duration_ms", time.Since(resolveStart).Milliseconds())

		queryStart := time.Now()
		gitInfo, err := queryGit(ctx, workdir)
		queryDur := time.Since(queryStart)
		if err != nil {
			p.logger.Warn("git query failed", "workdir", workdir, "duration_ms", queryDur.Milliseconds(), "error", err)
			continue
		}
		if gitInfo == nil {
			continue
		}
		p.logger.Debug("git query", "workdir", workdir, "branch", gitInfo.Branch, "duration_ms", queryDur.Milliseconds())

		filesStart := time.Now()
		changedFiles, err := queryChangedFiles(ctx, workdir)
		filesDur := time.Since(filesStart)
		if err != nil {
			p.logger.Warn("git changed files query failed", "workdir", workdir, "duration_ms", filesDur.Milliseconds(), "error", err)
			changedFiles = nil
		} else {
			p.logger.Debug("git changed files", "workdir", workdir, "file_count", len(changedFiles), "duration_ms", filesDur.Milliseconds())
		}

		// Look up PR number for non-default branches
		prNumber := 0
		if gitInfo.RemoteURL != nil && gitInfo.Branch != "main" && gitInfo.Branch != "master" && gitInfo.Branch != "HEAD" {
			prNumber = p.lookupPR(ctx, *gitInfo.RemoteURL, gitInfo.Branch, workdir)
		}

		dbStart := time.Now()
		for _, agent := range dirAgents {
			sessionID := agent.SessionID
			var sidPtr *string
			if sessionID != "" {
				sidPtr = &sessionID
			}

			snap := &store.GitSnapshot{
				AgentName:        agent.AgentName,
				AgentType:        agent.AgentType,
				WorkingDirectory: workdir,
				Branch:           gitInfo.Branch,
				CommitHash:       gitInfo.CommitHash,
				CommitSubject:    gitInfo.CommitSubject,
				CommitTimestamp:  &gitInfo.CommitTimestamp,
				SessionID:        sidPtr,
				RemoteURL:        gitInfo.RemoteURL,
				PRNumber:         prNumber,
			}
			if err := p.store.UpsertGitSnapshot(ctx, snap); err != nil {
				p.logger.Warn("upsert snapshot failed", "agent", agent.AgentName, "error", err)
			}

			if changedFiles != nil {
				// Cap file count to prevent DB thrashing on large monorepos
				if len(changedFiles) > 2000 {
					p.logger.Warn("capping changed files", "agent", agent.AgentName, "total", len(changedFiles), "cap", 2000)
					changedFiles = changedFiles[:2000]
				}
				if err := p.store.ReplaceChangedFiles(ctx, agent.AgentName, workdir, changedFiles, sidPtr); err != nil {
					p.logger.Warn("replace changed files failed", "agent", agent.AgentName, "error", err)
				}
			}
		}
		p.logger.Debug("git db writes", "workdir", workdir, "agents", len(dirAgents), "duration_ms", time.Since(dbStart).Milliseconds())

		dirDur := time.Since(dirStart)
		if dirDur > 2*time.Second {
			p.logger.Warn("git poll slow for directory", "workdir", workdir, "total_ms", dirDur.Milliseconds(),
				"query_ms", queryDur.Milliseconds(), "files_ms", filesDur.Milliseconds())
		}
	}

	totalDur := time.Since(pollStart)
	p.logger.Debug("git poll complete", "dirs", len(dirToAgents), "total_ms", totalDur.Milliseconds())
	if totalDur > 5*time.Second {
		p.logger.Warn("git poll slow", "total_ms", totalDur.Milliseconds(), "dirs", len(dirToAgents))
	}

	return nil
}

func (p *GitPoller) discoverAgents(ctx context.Context) ([]AgentInfo, error) {
	if p.discoverFn != nil {
		return p.discoverFn(ctx)
	}
	// Default: discover from runtime
	return p.runtime.ListAgents(ctx)
}


// gitInfo holds parsed git query results.
type gitInfo struct {
	Branch          string
	CommitHash      string
	CommitSubject   string
	CommitTimestamp  string
	RemoteURL       *string
}

func queryGit(ctx context.Context, workdir string) (*gitInfo, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	logger := slog.Default().With("service", "git_poller", "workdir", workdir)

	// Get branch name
	t0 := time.Now()
	out, err := executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "rev-parse", "--abbrev-ref", "HEAD").Output()
	logger.Debug("git rev-parse", "duration_ms", time.Since(t0).Milliseconds())
	if err != nil {
		return nil, err
	}
	branch := strings.TrimSpace(string(out))

	// Get latest commit
	t1 := time.Now()
	out, err = executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "log", "-1", "--format=%H|%s|%aI").Output()
	logger.Debug("git log", "duration_ms", time.Since(t1).Milliseconds())
	if err != nil {
		return nil, err
	}
	parts := strings.SplitN(strings.TrimSpace(string(out)), "|", 3)
	if len(parts) < 3 {
		return nil, fmt.Errorf("unexpected git log format")
	}

	// Get remote URL (best-effort)
	t2 := time.Now()
	var remoteURL *string
	out, err = executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "remote", "get-url", "origin").Output()
	logger.Debug("git remote", "duration_ms", time.Since(t2).Milliseconds())
	if err == nil {
		s := strings.TrimSpace(string(out))
		if s != "" {
			remoteURL = &s
		}
	}

	return &gitInfo{
		Branch:         branch,
		CommitHash:     parts[0],
		CommitSubject:  parts[1],
		CommitTimestamp: parts[2],
		RemoteURL:      remoteURL,
	}, nil
}

func queryChangedFiles(ctx context.Context, workdir string) ([]store.ChangedFile, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	logger := slog.Default().With("service", "git_poller", "workdir", workdir)

	t0 := time.Now()
	base := gitutil.GetDiffBase(ctx, workdir, "") // uses default branch_point mode
	logger.Debug("git diff base", "base", base, "duration_ms", time.Since(t0).Milliseconds())

	t1 := time.Now()
	baseTS := getBaseTimestamp(ctx, workdir, base)
	logger.Debug("git base timestamp", "duration_ms", time.Since(t1).Milliseconds())

	fileMap := make(map[string]store.ChangedFile)

	// git diff base --numstat
	t2 := time.Now()
	out, err := executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "diff", base, "--numstat").Output()
	if err == nil && len(out) > 0 {
		for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
			line = strings.TrimSpace(line)
			if line == "" {
				continue
			}
			parts := strings.SplitN(line, "\t", 3)
			if len(parts) < 3 {
				continue
			}
			a, _ := strconv.Atoi(parts[0])
			d, _ := strconv.Atoi(parts[1])
			fileMap[parts[2]] = store.ChangedFile{
				Filepath:  parts[2],
				Additions: a,
				Deletions: d,
				Status:    "M",
			}
		}
	}

	logger.Debug("git diff --numstat", "files", len(fileMap), "duration_ms", time.Since(t2).Milliseconds())

	// git status --porcelain for untracked files
	t3 := time.Now()
	out, err = executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "status", "--porcelain", "--untracked-files=normal").Output()
	if err == nil && len(out) > 0 {
		for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
			if len(line) < 4 {
				continue
			}
			statusCode := strings.TrimSpace(line[:2])
			fp := line[3:]
			if idx := strings.Index(fp, " -> "); idx >= 0 {
				fp = fp[idx+4:]
			}
			if statusCode == "??" {
				if _, exists := fileMap[fp]; !exists {
					fullPath := filepath.Join(workdir, fp)
					info, err := os.Stat(fullPath)
					if err != nil {
						continue
					}
					if info.ModTime().Unix() < int64(baseTS) {
						continue
					}
					fileMap[fp] = store.ChangedFile{
						Filepath:  fp,
						Additions: 0,
						Deletions: 0,
						Status:    "??",
					}
				}
			} else if _, exists := fileMap[fp]; exists {
				f := fileMap[fp]
				f.Status = statusCode
				fileMap[fp] = f
			}
		}
	}

	logger.Debug("git status --porcelain", "duration_ms", time.Since(t3).Milliseconds())

	files := make([]store.ChangedFile, 0, len(fileMap))
	for _, f := range fileMap {
		files = append(files, f)
	}

	if len(files) > 5000 {
		logger.Warn("large changed file set", "file_count", len(files), "base", base)
	}

	return files, nil
}

func getBaseTimestamp(ctx context.Context, workdir, baseRef string) float64 {
	out, err := executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "log", "-1", "--format=%ct", baseRef).Output()
	if err != nil {
		return 0
	}
	ts, err := strconv.ParseFloat(strings.TrimSpace(string(out)), 64)
	if err != nil {
		return 0
	}
	return ts
}

// lookupPR returns the PR number for the given remote+branch, using a cache
// keyed by remote_url::branch. The cache TTL matches the poller interval so
// each branch is looked up at most once per poll cycle.
func (p *GitPoller) lookupPR(ctx context.Context, remoteURL, branch, workdir string) int {
	cacheKey := remoteURL + "::" + branch
	if entry, ok := p.prCache[cacheKey]; ok && time.Now().Before(entry.expiresAt) {
		return entry.prNumber
	}

	prNumber := queryPRNumber(ctx, branch, workdir, p.logger)
	p.prCache[cacheKey] = prCacheEntry{
		prNumber:  prNumber,
		expiresAt: time.Now().Add(p.interval),
	}
	return prNumber
}

// queryPRNumber shells out to `gh pr list` to find an open or merged PR for a branch.
func queryPRNumber(ctx context.Context, branch, workdir string, logger *slog.Logger) int {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	out, err := executil.Command(ctx, "gh", "pr", "list",
		"--head", branch,
		"--state", "all",
		"--json", "number",
		"--limit", "1",
		"--repo", getRepoFromWorkdir(ctx, workdir),
	).Output()
	if err != nil {
		logger.Debug("gh pr list failed", "branch", branch, "error", err)
		return 0
	}

	// Parse minimal JSON: [{"number":42}]
	s := strings.TrimSpace(string(out))
	if s == "" || s == "[]" || s == "null" {
		return 0
	}
	// Quick parse without importing encoding/json for a single int field
	idx := strings.Index(s, `"number":`)
	if idx < 0 {
		return 0
	}
	numStr := s[idx+len(`"number":`):]
	numStr = strings.TrimLeft(numStr, " ")
	end := strings.IndexAny(numStr, ",}")
	if end > 0 {
		numStr = numStr[:end]
	}
	n, _ := strconv.Atoi(strings.TrimSpace(numStr))
	return n
}

func getRepoFromWorkdir(ctx context.Context, workdir string) string {
	out, err := executil.Command(ctx, "git", "--no-optional-locks", "-C", workdir, "remote", "get-url", "origin").Output()
	if err != nil {
		return ""
	}
	remote := strings.TrimSpace(string(out))

	// Convert git@github.com:org/repo.git or https://github.com/org/repo.git to org/repo
	remote = strings.TrimSuffix(remote, ".git")
	if strings.HasPrefix(remote, "git@") {
		// git@github.com:org/repo
		parts := strings.SplitN(remote, ":", 2)
		if len(parts) == 2 {
			return parts[1]
		}
	}
	// https://github.com/org/repo
	if idx := strings.Index(remote, "github.com/"); idx >= 0 {
		return remote[idx+len("github.com/"):]
	}
	return remote
}
